srclimit.c

```diff
@@ -136,3 +188,241 @@ srclimit_done(int id)
+
+	if (!penalty_cfg.enabled)
+		return;
+	if (penalty_exempt != NULL) {
+		if (addr_ntop(addr, addrnetmask, sizeof(addrnetmask)) != 0)
+			return; /* shouldn't happen */
+		if (addr_match_list(addrnetmask, penalty_exempt) == 1) {
+			debug3_f("address %s is exempt", addrnetmask);
+			return;
+		}
+	}
+
+	switch (penalty_type) {
+	case SRCLIMIT_PENALTY_NONE:
+		return;
+	case SRCLIMIT_PENALTY_CRASH:
+		penalty_secs = penalty_cfg.penalty_crash;
+		reason = "penalty: caused crash";
+		break;
+	case SRCLIMIT_PENALTY_AUTHFAIL:
+		penalty_secs = penalty_cfg.penalty_authfail;
+		reason = "penalty: failed authentication";
+		break;
+	case SRCLIMIT_PENALTY_NOAUTH:
+		penalty_secs = penalty_cfg.penalty_noauth;
+		reason = "penalty: connections without attempting authentication";
+		break;
+	case SRCLIMIT_PENALTY_GRACE_EXCEEDED:
+		penalty_secs = penalty_cfg.penalty_crash;
+		reason = "penalty: exceeded LoginGraceTime";
+		break;
+	default:
+		fatal_f("internal error: unknown penalty %d", penalty_type);
+	}
+	bits = addr->af == AF_INET ? ipv4_masklen : ipv6_masklen;
+	if (srclimit_mask_addr(addr, bits, &masked) != 0)
+		return;
+	addr_masklen_ntop(addr, bits, addrnetmask, sizeof(addrnetmask));
+
+	now = monotime();
+	expire_penalties(now);
+	if (npenalties > (size_t)penalty_cfg.max_sources &&
+	    penalty_cfg.overflow_mode == PER_SOURCE_PENALTY_OVERFLOW_DENY_ALL) {
+		verbose_f("penalty table full, cannot penalise %s for %s",
+		    addrnetmask, reason);
+		return;
+	}
+
+	penalty = xcalloc(1, sizeof(*penalty));
+	penalty->addr = masked;
+	penalty->expiry = now + penalty_secs;
+	penalty->reason = reason;
+	if ((existing = RB_INSERT(penalties_by_addr, &penalties_by_addr,
+	    penalty)) == NULL) {
+		/* penalty didn't previously exist */
+		if (penalty_secs > penalty_cfg.penalty_min)
+			penalty->active = 1;
+		if (RB_INSERT(penalties_by_expiry, &penalties_by_expiry,
+		    penalty) != NULL)
+			fatal_f("internal error: penalty tables corrupt");
+		verbose_f("%s: new %s penalty of %d seconds for %s",
+		    addrnetmask, penalty->active ? "active" : "deferred",
+		    penalty_secs, reason);
+		if (++npenalties > (size_t)penalty_cfg.max_sources)
+			srclimit_remove_penalty(); /* permissive */
+		return;
+	}
+	/* An entry already existed. Accumulate penalty up to maximum */
+	debug_f("%s penalty for %s already exists, %lld seconds remaining",
+	    existing->active ? "active" : "inactive",
+	    addrnetmask, (long long)(existing->expiry - now));
+	existing->expiry += penalty_secs;
+	if (existing->expiry - now > penalty_cfg.penalty_max)
+		existing->expiry = now + penalty_cfg.penalty_max;
+	if (existing->expiry - now > penalty_cfg.penalty_min &&
+	    !existing->active) {
+		verbose_f("%s: activating penalty of %lld seconds for %s",
+		    addrnetmask, (long long)(existing->expiry - now), reason);
+		existing->active = 1;
+	}
+	existing->reason = penalty->reason;
+	free(penalty);
+	/* Expiry has changed, re-insert into expiry tree */
```

Comment:
I think you have do:
1) remove 2) change 3) reinsert
since the tree is inconsistent as soon as you change the attrib used in the compare operation

Reply:
oops, done